torch/export/export_schema.py

```diff
@@ -0,0 +1,326 @@
+#     MemoryFormat?
+#     Layout?
+#     Device?
+
+
+################################################################################
+# Following section is the defining the schema of serializing a concrete tensor
+
+# TensorMeta is a decription of a tensor, without the actual data (,effectively maps to FakeTensor)
+# TensorMeta has multliple uses
+#   1. Represent the property of a concrete tensor backed by a storage
+#     - This is used in the serialization of a concrete tensor, e.g. model weight
+#     - In this case, sizes and strides must be concrete ints, and cannot be symbolic
+#     - stride and storage_offset have to used to correctly reconstruct the tensor from the storage
+#   2. Represent the property of a virtual tensor (see IValue below)
+#     - In this case, sizes and strides can be either concrete ints or symbolic ints.
+#     - device/strides/storage_offset/layout/memory_format are tied to pytorch's implementation.
+#       These are faithful capture of tensor's detail in pytorch's executions during tracing
+#       However, it's up to downstream system on how to utilized these fields
+#       In another word, these feilds are suggestive, rather than mandatory.
+
+class TensorMeta:
+    dtype: ScalarType
+    sizes: List[SymInt]
+
+    # needed for training
+    requires_grad: bool
+
+    # !!! see description above, there are subtle difference on how these fields should be interpreted
+    device: Device
+    strides: List[SymInt]
+    storage_offset: int
+    layout: Layout
+    memory_format: MemoryFormat
+
+
+@dataclass
+class Buffer:
+    buffer: bytes
+
+
+@dataclass
+class ExternalBuffer:
+    # !!! Example from onnx
+    #   // Data can be stored inside the protobuf file using type-specific fields or raw_data.
+    #   // Alternatively, raw bytes data can be stored in an external file, using the external_data field.
+    #   // external_data stores key-value pairs describing data location. Recognized keys are:
+    #   // - "location" (required) - POSIX filesystem path relative to the directory where the ONNX
+    #   //                           protobuf model was stored
+    #   // - "offset" (optional) - position of byte at which stored data begins. Integer stored as string.
+    #   //                         Offset values SHOULD be multiples 4096 (page size) to enable mmap support.
+    #   // - "length" (optional) - number of bytes containing data. Integer stored as string.
+    #   // - "checksum" (optional) - SHA1 digest of file specified in under 'location' key.
+    #   repeated StringStringEntryProto external_data
+    location: str
+    offset: str     # !!! Consider using int, but int has int_max limitation
+    length: str     # !!! Consider using int, but int has int_max limitation
+    checksum: str
+
+
+@dataclass
+class Storage:
+    class DataLocation(Enum):
+        Default = auto()
+        External = auto()
+
+    data_location: DataLocation
+    data: Union[Buffer, ExternalBuffer]
```

Comment:
You are trying to make the Tensor lazily loadable? Why is this part of schema though

Reply:
Ditto. 

We should supports both weights in model file, as well as in an external file.

Reply:
To be more specific, the **program** file should contain what things external environment supply. This includes: 1. set of operators and their schema, and 2. set of named tensors and their shape/dtype/metadata (i.e. FakeTensor). 

The thing that loads a program will get a graph, which is equivalent of the "code" part of the model (no weights). Then then, the weights are loaded separately from a different file / network / initialized with random whatever, as part of initializing an **environment**.

When the graph is executed or manipulated, when a weight is being asked (in get_attr, nodes) we look in the environment to get what the actual value are.